### PR TITLE
Add support for dual stack IPv6

### LINKED ISSUE #2554: Add IPv6 dual stack support
Upstream is adding support to CRI for IPv6. We need to support this in CRI-O for 1.16.

### PR DESCRIPTION
**Pull request by saschagrunert**
We have to vendor the latest Kubernetes master as well as OCICNI to get
the latest support for dual-stack IPv6. This adds an `AdditionalIps`
field to the network status, which will be internally handled as a
simple slice of strings. This means that the pod annotations now can
contain multiple IPs as well. The same applied to the inspect HTTP API.

Relates to: https://github.com/kubernetes/kubernetes/pull/73977
Current status: Under testing since the kubernetes PR has been merged. 

Closes #2554

### CODE REVIEW

**Comment by uablrek** [comment]
No, the (new) `podIPs` array in the pod object still contains only the ipv4 address;
```
# kubectl get pod alpine-deployment-568f6756d7-j8wqt -o json | jq .stat
us.podIPs
[
  {
    "ip": "11.0.2.2"
  }
]
```
I applied the PR with;
```
curl -L https://github.com/cri-o/cri-o/pull/2447.patch | patch -p1
```
On target;
```
# crio -v
crio version 1.14.0-dev
commit: "f12dcd768411cc16fe399e36d2be047df1e51abd-dirty"
```
Here is the log from `kubelet`;
```
I0604 15:50:48.467110     231 kuberuntime_manager.go:943] updating runtime config through cri with podcidr 11.0.0.0/24,1100::/24
I0604 15:50:48.468392     231 kubelet_network.go:77] Setting Pod CIDR:  -> 11.0.0.0/24,1100::/24
I0604 15:50:49.101298     231 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "default-token-mbcvt" (UniqueName: "kubernetes.io/secret/120e9e62-6ecb-43c6-897d-bb1e7c9e9c7a-default-token-mbcvt") pod "coredns-78d6576697-lcqxb" (UID: "120e9e62-6ecb-43c6-897d-bb1e7c9e9c7a") 
I0604 15:50:49.101355     231 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "config-volume" (UniqueName: "kubernetes.io/configmap/120e9e62-6ecb-43c6-897d-bb1e7c9e9c7a-config-volume") pod "coredns-78d6576697-lcqxb" (UID: "120e9e62-6ecb-43c6-897d-bb1e7c9e9c7a") 
W0604 15:50:51.138900     231 kuberuntime_sandbox.go:250] Pod Sandbox reported an unparseable IP 11.0.0.2
W0604 15:50:51.188404     231 kuberuntime_sandbox.go:250] Pod Sandbox reported an unparseable IP 11.0.0.2
W0604 15:50:52.192514     231 kuberuntime_sandbox.go:250] Pod Sandbox reported an unparseable IP 11.0.0.2
I0604 15:52:04.760083     231 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "default-token-mbcvt" (UniqueName: "kubernetes.io/secret/ae2b7d2c-703f-48f6-99b9-a31d5f5ac38e-default-token-mbcvt") pod "alpine-deployment-568f6756d7-pr6hx" (UID: "ae2b7d2c-703f-48f6-99b9-a31d5f5ac38e") 
W0604 15:52:06.291980     231 kuberuntime_sandbox.go:250] Pod Sandbox reported an unparseable IP 11.0.0.3
W0604 15:52:06.494706     231 kuberuntime_sandbox.go:250] Pod Sandbox reported an unparseable IP 11.0.0.3
W0604 15:52:07.496522     231 kuberuntime_sandbox.go:250] Pod Sandbox reported an unparseable IP 11.0.0.3
```
And what I hope is relevant in the cri-o log (it is a lot of things in it);
```
time="2019-06-04 15:50:51.138735986Z" level=debug msg="PodSandboxStatusResponse: &PodSandboxStatusResponse{Status:&PodSandboxStatus{Id:94ef37d9fce1f38c2a10569bf8e23828385af8a
629bc012a06fd78ea5a6ac588,Metadata:&PodSandboxMetadata{Name:coredns-78d6576697-lcqxb,Uid:120e9e62-6ecb-43c6-897d-bb1e7c9e9c7a,Namespace:default,Attempt:0,},State:SANDBOX_READ
Y,CreatedAt:1559663449523493978,Network:&PodSandboxNetworkStatus{Ip:11.0.0.2,},Linux:&LinuxPodSandboxStatus{Namespaces:&Namespace{Options:&NamespaceOption{Network:POD,Pid:CON
TAINER,Ipc:POD,},},},Labels:map[string]string{app: coredns,io.kubernetes.container.name: POD,io.kubernetes.pod.name: coredns-78d6576697-lcqxb,io.kubernetes.pod.namespace: def
ault,io.kubernetes.pod.uid: 120e9e62-6ecb-43c6-897d-bb1e7c9e9c7a,pod-template-hash: 78d6576697,},Annotations:map[string]string{kubernetes.io/config.seen: 2019-06-04T15:50:49.
079194029Z,kubernetes.io/config.source: api,},RuntimeHandler:,},Info:map[string]string{},}"
```
The `Ip:11.0.0.2,` looks like the beginning of a list, but no ipv6 address in.

**Comment by saschagrunert** [comment]
Thanks for the input, I guess we have to adapt code within [OCI CNI](https://github.com/cri-o/ocicni), too.

**Comment by uablrek** [comment]
Re-build/re-run. Logs at [crio.log.gz](https://github.com/cri-o/cri-o/files/3257596/crio.log.gz)

**Comment by saschagrunert** [comment]
> Re-build/re-run. Logs at [crio.log.gz](https://github.com/cri-o/cri-o/files/3257596/crio.log.gz)

The logs are looking good so far. I see both interfaces discovered. I guess we have to rebase to the K8S PR to make a test run actually work.

**Comment by uablrek** [comment]
Latest run ([crio.log.gz](https://github.com/cri-o/cri-o/files/3261127/crio.log.gz);
```
vm-002 ~ # kubectl get pod $p -o json | jq .status.podIPs
[
  {
    "ip": "1100:100::2"
  }
]
```
Now only the ipv6 address is seen.

**Comment by uablrek** [comment]
```
  "hostIP": "192.168.1.3",
  "phase": "Running",
  "podIP": "1100:100::2",
  "podIPs": [
    {
      "ip": "1100:100::2"
    }
  ],
```

**Comment by saschagrunert** [comment]
I changed the implementation that now the `Ips` are set within the `PodSandboxNetworkStatus`. For now we also set the old `Ip` field, whereas the discussion does not seem resolved on the [related PR](https://github.com/kubernetes/kubernetes/pull/73977/files#r291056284).

**Comment by uablrek** [comment]
Thanks, I works perfectly now;
```
# kubectl get pod $p -o json | jq .status.podIPs
[
  {
    "ip": "11.0.0.3"
  },
  {
    "ip": "1100::3"
  }
]
```

**Comment by saschagrunert** [comment]
We need to adapt libpod before actually having this to work: https://github.com/containers/libpod/pull/3486

**Comment by saschagrunert** [comment]
It looks like that we have to update the CNI plugins on the machines to make the tests work.

**Comment by saschagrunert** [comment]
@giuseppe , for fedora it works (see the integration test). The RHEL AMI build failed so we have to wait for a fix about that.

**Comment by saschagrunert** [comment]
To fix the RHEL builds we have to wait for support at https://github.com/openshift/release/issues/4363

**Comment by saschagrunert** [comment]
Finally, PTAL @haircommander @mrunalp @rhatdan and @vrothberg

**Comment by rhatdan** [comment]
LGTM

File: internal/lib/sandbox/sandbox.go [hunk 1]
```diff
@@ -186,8 +186,8 @@ func (s *Sandbox) SeccompProfilePath() string {
 }
 
 // AddIP stores the ip in the sandbox
```
Comment (vrothberg) [inline internal/lib/sandbox/sandbox.go]: Absolutely non-blocking: I wonder why the linter doesn't complain about the missing 's'.

File: internal/lib/sandbox/sandbox.go [hunk 2]
```diff
@@ -206,8 +206,8 @@ func (s *Sandbox) StopMutex() *sync.RWMutex {
 }
 
 // IP returns the ip of the sandbox
-func (s *Sandbox) IP() string {
```
Comment (vrothberg) [inline internal/lib/sandbox/sandbox.go]: Same here.

File: server/sandbox_network.go [hunk 3]
```diff
@@ -54,51 +54,60 @@ func (s *Server) networkStart(sb *sandbox.Sandbox) (podIP string, result cnitype
...
+			}
+
+			err = s.hostportManager.Add(sb.ID(), &hostport.PodPortMapping{
+				Name:         sb.Name(),
+				PortMappings: sb.PortMappings(),
+				IP:           ip,
+				HostNetwork:  false,
+			}, "lo")
+			if err != nil {
+				err = fmt.Errorf("failed to add hostport mapping for sandbox %s(%s): %v", sb.Name(), sb.ID(), err)
```
Comment (TomSweeneyRedHat) [inline server/sandbox_network.go]: Do we need to add the `ip` field to this message?   Perhaps  `... hostport mapping using IP %v for sandbox ...`

**Comment by TomSweeneyRedHat** [comment]
One small error comment that can be addressed later if necessary, otherwise LGTM

**Comment by saschagrunert** [comment]
> One small error comment that can be addressed later if necessary, otherwise LGTM

Okay thanks, I will fix it either later or within this PR, depending on the CI situation :)

**Comment by TomSweeneyRedHat** [comment]
@saschagrunert yes, definitely if the CI lets this one merge, please do so.

**Comment by saschagrunert** [comment]
Oh, no kata seems broken, I will check if I can fix it on upstream.

**Comment by saschagrunert** [comment]
I try to tackle the kata issue with https://github.com/kata-containers/tests/pull/1816

**Comment by saschagrunert** [comment]
Let's see if we can get this baby in...

/retest

**Comment by saschagrunert** [comment]
I had to rebase to resolve the conflicts, PTAL :innocent:

**Pull request merged**